Lib/concurrent/futures/_base.py

```diff
@@ -544,18 +545,40 @@ def map(self, fn, *iterables, timeout=None, chunksize=1):
         """
         if timeout is not None:
             end_time = timeout + time.time()
+        if prefetch is None:
+            prefetch = self._max_workers
+        if prefetch < 0:
+            raise ValueError("prefetch count may not be negative")
 
-        fs = [self.submit(fn, *args) for args in zip(*iterables)]
+        argsiter = zip(*iterables)
+
+        fs = collections.deque(self.submit(fn, *args) for args in itertools.islice(argsiter, self._max_workers + prefetch))
 
         # Yield must be hidden in closure so that the futures are submitted
         # before the first iterator value is required.
         def result_iterator():
+            nonlocal argsiter
             try:
-                for future in fs:
+                while fs:
                     if timeout is None:
-                        yield future.result()
+                        res = fs[0].result()
                     else:
-                        yield future.result(end_time - time.time())
+                        res = fs[0].result(end_time - time.time())
+
+                    # Got a result, future needn't be cancelled
+                    del fs[0]
+
```

Comment:
Oh I didn't mean to change the semantics or the surrounding code. I just meant only for this single line: you can replace `del fs[0]` with `fs.popleft()` (either way you're deleting the leftmost item).